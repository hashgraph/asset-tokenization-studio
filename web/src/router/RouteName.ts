export enum RouteName {
  Landing = "LANDING",
  Dashboard = "DASHBOARD",
  AddSecurity = "ADD_SECURITY",
  CreateSecurity = "CREATE_SECURITY",
  CreateEquity = "CREATE_EQUITY",
  CreateBond = "CREATE_BOND",
  DigitalSecurityDetails = "DIGITAL_SECURITY_DETAILS",
  DigitalSecuritiesList = "DIGITAL_SECURITIES_LIST",
  DigitalSecurityMint = "DIGITAL_SECURITY_MINT",
  DigitalSecurityTransfer = "DIGITAL_SECURITY_TRANSFER",
  DigitalSecurityForceTransfer = "DIGITAL_SECURITY_FORCE_TRANSFER",
  DigitalSecurityRedeem = "DIGITAL_SECURITY_REDEEM",
  DigitalSecurityForceRedeem = "DIGITAL_SECURITY_FORCE_REDEEM",
}
