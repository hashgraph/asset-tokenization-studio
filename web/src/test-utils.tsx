import React from "react";
import { ChakraProvider } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render, waitFor } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import { I18nextProvider } from "react-i18next";
import { Router } from "react-router-dom";
import i18n from "./i18n";
import theme from "./theme";
import userEvent from "@testing-library/user-event";

export const selectCalendar = async (
  component: RenderResult,
  id: string,
  day: string | number = new Date().getDate(),
) => {
  const calendar = component.getByTestId(id);
  await userEvent.click(calendar);

  await waitFor(() => {
    const daysToSelect = component.getAllByTestId(`day-${day}`);

    const dayToSelect = daysToSelect.find(
      (day) => !day.hasAttribute("disabled"),
    );

    if (dayToSelect) {
      userEvent.click(dayToSelect);
    }
  });

  await userEvent.click(document.body);
};

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: () => {},
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory,
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <I18nextProvider i18n={i18n}>
      <QueryClientProvider client={testQueryClient}>
        <ChakraProvider theme={theme}>
          <Router location={history.location} navigator={history}>
            {children}
          </Router>
        </ChakraProvider>
      </QueryClientProvider>
    </I18nextProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history,
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {},
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => (
      <AllProviders history={history}>{children}</AllProviders>
    ),
    ...options,
  });

export { customRender as render };
