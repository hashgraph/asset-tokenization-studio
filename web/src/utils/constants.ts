export const METAMASK_URL = "https://metamask.io/download/";

export enum MetamaskStatus {
  connected = "CONNECTED",
  connecting = "CONNECTING",
  disconnected = "DISCONNECTED",
  uninstalled = "UNINSTALLED",
}

export enum User {
  admin = "admin",
  holder = "holder",
  general = "general",
}

export const LOCALE = "en-US";
export const COUPONS_FACTOR = 1000;
export const NOMINAL_VALUE_FACTOR = 100;
