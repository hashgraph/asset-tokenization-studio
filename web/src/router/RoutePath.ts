import { RouteName } from "./RouteName";

export const RoutePath: Record<RouteName, string> = {
  [RouteName.Dashboard]: "/",
  [RouteName.Landing]: "/connect-to-metamask",
  [RouteName.DigitalSecurityDetails]: "/security/:id",
  [RouteName.DigitalSecurityMint]: "/security/:id/mint",
  [RouteName.DigitalSecurityTransfer]: "/security/:id/transfer",
  [RouteName.DigitalSecurityForceTransfer]: "/security/:id/forceTransfer",
  [RouteName.DigitalSecurityRedeem]: "/security/:id/redeem",
  [RouteName.DigitalSecurityForceRedeem]: "/security/:id/forceRedeem",
  [RouteName.DigitalSecuritiesList]: "/list/:type",
  [RouteName.AddSecurity]: "/security/add",
  [RouteName.CreateSecurity]: "/security/create",
  [RouteName.CreateEquity]: "/security/create/equity",
  [RouteName.CreateBond]: "/security/create/bond",
};
