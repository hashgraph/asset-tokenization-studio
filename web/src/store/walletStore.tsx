import { create } from "zustand";
import { WalletStatus } from "../utils/constants";
import type { InitializationData } from "@hashgraph/asset-tokenization-sdk";
import { NetworkData } from "@hashgraph/asset-tokenization-sdk";

type WalletStoreStatus =
  | WalletStatus.disconnected
  | WalletStatus.connected
  | WalletStatus.connecting
  | WalletStatus.uninstalled;

interface WalletStore {
  address: string;
  setAddress: (address: string) => void;
  connectionStatus: WalletStoreStatus;
  setConnectionStatus: (status: WalletStoreStatus) => void;
  reset: () => void;
  data: InitializationData | null;
  network: NetworkData | null;
  setPairedWallet: (data: InitializationData, network: NetworkData) => void;
}

export const useWalletStore = create<WalletStore>((set) => ({
  address: "",
  setAddress: (address: string) =>
    set((state: WalletStore) => ({
      ...state,
      address,
      connectionStatus: WalletStatus.connected,
    })),
  connectionStatus: WalletStatus.disconnected,
  setConnectionStatus: (status: WalletStoreStatus) =>
    set((state: WalletStore) => ({ ...state, connectionStatus: status })),
  reset: () =>
    set((state: WalletStore) => ({
      ...state,
      address: "",
      data: null,
      network: null,
      connectionStatus: WalletStatus.disconnected,
    })),
  data: null,
  network: null,
  setPairedWallet: (data: InitializationData, network: NetworkData) =>
    set((state: WalletStore) => ({
      ...state,
      data,
      network,
      address: data.account?.id.value,
      connectionStatus: WalletStatus.connected,
    })),
}));
