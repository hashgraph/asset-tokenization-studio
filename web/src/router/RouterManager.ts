import type { ReverseParams } from "named-urls";
import { reverse } from "named-urls";
import { router } from ".";
import { RouteName } from "./RouteName";
import { RoutePath } from "./RoutePath";

export interface RouteParams {
  params?: ReverseParams;
  extra?: string;
  state?: object;
}

export class BaseRouterManager {
  constructor(private routes: Record<RouteName, string> = RoutePath) {}

  #getPath(name: RouteName, { params, extra = "" }: RouteParams) {
    const pattern = this.routes[name] + extra;
    return reverse(pattern, params);
  }

  to(name: RouteName, { state, ...params }: RouteParams = {}) {
    const pathname = this.#getPath(name, params);
    return router.navigate({ pathname }, { state });
  }

  getUrl(name: RouteName, params: RouteParams = {}) {
    const path = this.#getPath(name, params);
    return reverse(path);
  }

  goBack() {
    return router.navigate(-1);
  }

  goDashboard() {
    return this.to(RouteName.Dashboard);
  }
}

export const RouterManager = new BaseRouterManager(RoutePath);
