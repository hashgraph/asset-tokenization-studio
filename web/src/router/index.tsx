import { RouterProvider, createBrowserRouter } from "react-router-dom";
import { routes } from "./Routes";
import { useAccountStore } from "../store/accountStore";
import { useEffect } from "react";
import { useSecurityStore } from "../store/securityStore";

export const router = createBrowserRouter(routes);

const AppRouter = () => {
  const { isResetted, reset: resetAccountStore } = useAccountStore();
  const { reset: reserSecurityStore } = useSecurityStore();

  useEffect(() => {
    if (!isResetted) {
      resetAccountStore();
      reserSecurityStore();
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return <RouterProvider router={router} />;
};

export default AppRouter;
