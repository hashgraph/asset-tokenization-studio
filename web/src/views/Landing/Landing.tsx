import { useEffect } from "react";
import {
  Center,
  ListItem,
  Link,
  OrderedList,
  Stack,
  VStack,
  Box,
} from "@chakra-ui/react";
import { PhosphorIcon, Text } from "@hashgraph/securitytoken-uicomponents/Foundations";
import { Button } from "@hashgraph/securitytoken-uicomponents/Interaction";
import landingBackground from "../../assets/layer.png";
import { useWalletStore } from "../../store/walletStore";
import _capitalize from "lodash/capitalize";
import { Trans, useTranslation } from "react-i18next";
import { RouterManager } from "../../router/RouterManager";
import { RouteName } from "../../router/RouteName";
import { PopUp } from "@hashgraph/securitytoken-uicomponents";
import { Wallet } from "@phosphor-icons/react";
import { Weight } from "@hashgraph/securitytoken-uicomponents/Foundations";
import { MetamaskStatus, METAMASK_URL, User } from "../../utils/constants";
import { useWalletConnection } from "../../hooks/useWalletConnection";
import { useUserStore } from "../../store/userStore";

export const Landing = () => {
  const { t } = useTranslation("landing");
  const { t: tGlobals } = useTranslation("globals");
  const { t: tConnecting } = useTranslation("landing", {
    keyPrefix: "metamaskPopup.connecting",
  });
  const { t: tUninstalled } = useTranslation("landing", {
    keyPrefix: "metamaskPopup.uninstalled",
  });
  const { connectionStatus, reset } = useWalletStore();
  const { handleConnectWallet } = useWalletConnection();
  const { setType } = useUserStore();

  const handleInstallButton = () => {
    window.open(METAMASK_URL, "_blank");
    reset();
  };

  useEffect(() => {
    const connected = connectionStatus === MetamaskStatus.connected;
    if (connected) {
      RouterManager.to(RouteName.Dashboard);
    }

    const disconnected = connectionStatus === MetamaskStatus.disconnected;
    if (disconnected) {
      setType(User.general);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [connectionStatus]);

  const isLoading = connectionStatus === MetamaskStatus.connecting;
  if (isLoading) {
    return (
      <Center h="full" data-testid="connecting-to-metamask">
        <PopUp isOpen={true} onClose={() => reset()}>
          <PhosphorIcon
            as={Wallet}
            size="md"
            sx={{ color: "primary.500" }}
            weight={Weight.Fill}
          />
          <Text mt={2} textStyle="ElementsMediumMD">
            {tConnecting("title")}
          </Text>
          <Text mt={1} textStyle="ElementsRegularXS">
            {tConnecting("description")}
          </Text>
        </PopUp>
      </Center>
    );
  }

  const uninstalled = connectionStatus === MetamaskStatus.uninstalled;
  if (uninstalled) {
    return (
      <Center h="full" data-testid="install-metamask">
        <PopUp isOpen={true} onClose={() => reset()}>
          <PhosphorIcon
            as={Wallet}
            size="md"
            sx={{ color: "primary.500" }}
            weight={Weight.Fill}
          />
          <Text mt={2} textStyle="ElementsMediumMD">
            {tUninstalled("title")}
          </Text>
          <Text mt={1} textStyle="ElementsRegularXS" w="240px">
            {tUninstalled("description")}
          </Text>
          <Button
            data-testid="install-metamask-extension-button"
            size="md"
            onClick={() => handleInstallButton()}
            color="primary.500"
            mt={6}
          >
            {tUninstalled("button")}
          </Button>
        </PopUp>
      </Center>
    );
  }

  return (
    <Stack
      data-testid="landing-page"
      w="full"
      h="full"
      backgroundImage={`url(${landingBackground})`}
      backgroundSize="cover"
      backgroundPosition="center"
      backgroundRepeat="no-repeat"
      justifyContent="center"
    >
      <VStack alignItems="flex-start" marginLeft={150}>
        <Text
          data-testid="title"
          textStyle="ElementsSemiboldXL"
          lineHeight="57px"
          fontSize="55px"
          color="neutral.900"
          maxW="470px"
        >
          {`${t("connectYour")} `}
          <Text as="span" color="primary.500">
            {t("Wallet")}
          </Text>
        </Text>
        <Text
          data-testid="subtitle"
          textStyle="ElementsLightLG"
          width="329px"
          my={3}
        >
          {t("welcomeMessage")}
        </Text>
        <Box textStyle="ElementsLightSM">
          <Trans
            t={t}
            i18nKey="instructions"
            components={{
              p: <Text />,
              ol: <OrderedList />,
              li: <ListItem />,
              a: <Link isExternal sx={{ display: "inline" }} />,
            }}
          />
        </Box>

        <Button
          data-testid="connect-to-metamask-landing-button"
          onClick={() => handleConnectWallet()}
          mt={7}
        >
          <Text textStyle="ElementsMediumSM" color="neutral.650">
            {tGlobals("connectMetamask")}
          </Text>
        </Button>
      </VStack>
    </Stack>
  );
};
