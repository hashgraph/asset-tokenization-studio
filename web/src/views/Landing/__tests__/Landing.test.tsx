import { waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Landing } from "../Landing";
import { render } from "../../../test-utils";
import { RouterManager } from "../../../router/RouterManager";
import { RouteName } from "../../../router/RouteName";
import { useWalletStore } from "../../../store/walletStore";
import { METAMASK_URL, MetamaskStatus } from "../../../utils/constants";

jest.mock("../../../router/RouterManager", () => ({
  RouterManager: {
    ...jest.requireActual("../../../router/RouterManager").RouterManager,
    to: jest.fn(),
  },
}));

const initialStoreState = useWalletStore.getState();

describe(`${Landing.name}`, () => {
  const accounts = ["0xd8da6bf26964af9d7eed9e03e53415d37aa96045"];

  afterEach(() => {
    jest.clearAllMocks();
    useWalletStore.setState(initialStoreState, true);
  });

  test("should render correctly", async () => {
    const component = render(<Landing />);

    await waitFor(() => {
      expect(component.getByTestId("landing-page"));
    });

    expect(component.asFragment()).toMatchSnapshot("disconnected");
  });

  test("if user connect to metamask then should redirect to dashboard", async () => {
    useWalletStore.setState(
      {
        ...initialStoreState,
        connectionStatus: MetamaskStatus.connected,
        address: accounts[0],
      },
      true,
    );

    const component = render(<Landing />);

    await waitFor(() => {
      expect(component.getByTestId("landing-page"));
    });

    const button = component.getByTestId("connect-to-metamask-landing-button");
    userEvent.click(button);

    await waitFor(() => {
      expect(component.queryByTestId("landing-page")).not.toBeInTheDocument();
    });

    expect(component.asFragment()).toMatchSnapshot("connecting");
    expect(RouterManager.to).toHaveBeenCalledTimes(1);
    expect(RouterManager.to).toHaveBeenCalledWith(RouteName.Dashboard);
  });

  test("if metamask can't connect then should go back to landing page", async () => {
    const component = render(<Landing />);

    await waitFor(() => {
      expect(component.getByTestId("landing-page"));
    });

    const button = component.getByTestId("connect-to-metamask-landing-button");

    userEvent.click(button);
    useWalletStore.setState(
      { ...initialStoreState, connectionStatus: MetamaskStatus.connecting },
      true,
    );

    await waitFor(() => {
      expect(component.queryByTestId("landing-page")).not.toBeInTheDocument();
      expect(
        component.getByTestId("connecting-to-metamask"),
      ).toBeInTheDocument();
    });

    useWalletStore.setState(initialStoreState, true);

    expect(RouterManager.to).toHaveBeenCalledTimes(0);

    await waitFor(() => {
      expect(component.getByTestId("landing-page")).toBeInTheDocument();
    });
  });

  test("if metamask is not installed then should open popup with button to install", async () => {
    const component = render(<Landing />);
    const uninstalled = {
      ...initialStoreState,
      connectionStatus: MetamaskStatus.uninstalled,
    };
    useWalletStore.setState(uninstalled, true);

    await waitFor(() => {
      expect(component.getByTestId("install-metamask")).toBeInTheDocument();
    });

    expect(component.asFragment()).toMatchSnapshot("uninstalled");
  });

  test("if click on install metamask button should open metamask page and go back to landing page", async () => {
    const oldWindowOpen = window.open;
    window.open = jest.fn();
    const component = render(<Landing />);

    const uninstalled = {
      ...initialStoreState,
      connectionStatus: MetamaskStatus.uninstalled,
    };
    useWalletStore.setState(uninstalled, true);

    await waitFor(() => {
      expect(component.getByTestId("install-metamask")).toBeInTheDocument();
    });

    const installButton = component.getByTestId(
      "install-metamask-extension-button",
    );
    userEvent.click(installButton);

    await waitFor(() => {
      expect(global.window.open).toHaveBeenCalledWith(METAMASK_URL, "_blank");
      expect(component.getByTestId("landing-page"));
    });

    window.open = oldWindowOpen;
  });
});
