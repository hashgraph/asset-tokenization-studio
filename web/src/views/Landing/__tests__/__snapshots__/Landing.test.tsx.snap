// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing if metamask is not installed then should open popup with button to install: uninstalled 1`] = `
<DocumentFragment>
  <div
    class="css-l84sf8"
    data-testid="install-metamask"
  />
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`Landing if user connect to metamask then should redirect to dashboard: connecting 1`] = `
<DocumentFragment>
  <div
    class="css-l84sf8"
    data-testid="install-metamask"
  />
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`Landing should render correctly: disconnected 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-31iced"
    data-testid="landing-page"
  >
    <div
      class="chakra-stack css-1h7sl3e"
    >
      <p
        class="css-hwd6mb"
        data-testid="title"
      >
        Connect your 
        <span
          class="css-qa8l2m"
        >
          Wallet
        </span>
      </p>
      <p
        class="css-b26fyd"
        data-testid="subtitle"
      >
        Please, connect your wallet to start operating.
      </p>
      <div
        class="css-ehmgu2"
      >
        <div
          class="chakra-accordion css-1gtanqs"
        >
          <div
            class="chakra-accordion__item css-j7dq76"
          >
            <button
              aria-controls="accordion-panel-:r0:"
              aria-expanded="false"
              class="chakra-accordion__button css-uttm9k"
              data-index="0"
              id="accordion-button-:r0:"
              type="button"
            >
              <div
                class="css-1eziwv"
              >
                Show Metamask instructions
              </div>
              <svg
                aria-hidden="true"
                class="chakra-icon chakra-accordion__icon css-186l2rg"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              class="chakra-collapse"
              style="overflow: hidden; display: none; opacity: 0; height: 0px;"
            >
              <div
                aria-labelledby="accordion-button-:r0:"
                class="chakra-accordion__panel css-1hb5ode"
                id="accordion-panel-:r0:"
                role="region"
              >
                <p
                  class="css-0"
                >
                  Instructions to set up your Hedera account and your Metamask wallet:
                </p>
                 
                <ol
                  class="css-1dg6mvm"
                  role="list"
                >
                  <li
                    class="css-0"
                  >
                    Configure the Hedera Testnet Network in Metamask and connect Metamask to it. Recommendation: 
                    <a
                      class="chakra-link css-1j7tat"
                      href="https://chainlist.org/chain/296"
                      rel="noopener"
                      target="_blank"
                    >
                      use Chainlist
                    </a>
                    .
                  </li>
                  <li
                    class="css-0"
                  >
                     Create an ECDSA account in the 
                    <a
                      class="chakra-link css-1j7tat"
                      href="https://portal.hedera.com/"
                      rel="noopener"
                      target="_blank"
                    >
                      Hedera Portal
                    </a>
                    .
                  </li>
                  <li
                    class="css-0"
                  >
                    Import the Private Key in Metamask (
                    <a
                      class="chakra-link css-1j7tat"
                      href="https://support.metamask.io/hc/en-us/articles/360015489331-How-to-import-an-account"
                      rel="noopener"
                      target="_blank"
                    >
                      instructions
                    </a>
                    ). 
                  </li>
                  <li
                    class="css-0"
                  >
                    Switch to that new account in Metamask (
                    <a
                      class="chakra-link css-1j7tat"
                      href="https://support.metamask.io/hc/en-us/articles/360061346311-Switching-accounts-in-MetaMask"
                      rel="noopener"
                      target="_blank"
                    >
                      instructions
                    </a>
                    ). 
                  </li>
                  <li
                    class="css-0"
                  >
                    Authorize the Token Studio site to use the account.
                  </li>
                </ol>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="chakra-stack css-ph6hc8"
      >
        <button
          class="chakra-button css-1145i8e"
          data-testid="connect-to-metamask-landing-button"
          type="button"
        >
          <p
            class="css-i0u7lf"
          >
            Connect Metamask
          </p>
        </button>
        <button
          class="chakra-button css-1145i8e"
          data-testid="connect-to-hwc-landing-button"
          type="button"
        >
          <p
            class="css-i0u7lf"
          >
            Connect Wallet Connect
          </p>
        </button>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
